presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -438,10 +440,59 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                             .unordered(true)
                             .build();
                 case FULL:
-                    // We can't say anything about the partitioning scheme because any partition of
-                    // a hash-partitioned join can produce nulls in case of a lack of matches
+                    // the join keys might not be output by the join node, so we need to see if we can express the partitioning scheme
+                    // for the join in terms of its outputs
+                    probeProperties = probeProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+                    buildProperties = buildProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+
+                    if (!SystemSessionProperties.isDeriveFullJoinProperties(session) ||
+                            !probeProperties.getNodePartitioning().isPresent() ||
+                            !buildProperties.getNodePartitioning().isPresent()) {
+                        return ActualProperties.builder()
+                                .global(probeProperties.isSingleNode() ? singleStreamPartition() : arbitraryPartition())
+                                .build();
+                    }
+
+                    checkState(
+                            probeProperties.getNodePartitioning().get().getArguments().size() == buildProperties.getNodePartitioning().get().getArguments().size(),
+                            "Inconsistent partitioning for left and right side of join: %s vs %s",
+                            probeProperties.getNodePartitioning().get(),
+                            buildProperties.getNodePartitioning().get());
+
+                    List<Partitioning.ArgumentBinding> newBindings = new ArrayList<>();
+                    for (int i = 0; i < probeProperties.getNodePartitioning().get().getArguments().size(); i++) {
+                        Partitioning.ArgumentBinding left = probeProperties.getNodePartitioning().get().getArguments().get(i);
```

Comment:
Is it obvious that partitioning symbols are in the same order as equi-join clauses?